### PR TITLE
Don't use the unreadChildren property of message threads as the unread count

### LINKED ISSUE #217: Unread emails count shows wrong number
**OS**: ArchLinux 64 bit
**Birdtray version**: 1.7.0
**Thunderbird version**: 68.3.0
**Birdtray origin**: ArchLinux's AUR
**Qt version**: _NA_

**Description**
The tray icon shows 4 unread messages event when I have no unread messages. I traced down the problem to the fact that I have 4 email accounts in Thunderbird, and when I configure BirdTray to monitor one of the Inbox folders then the unread count increases. This has not a 1:1 relationship, i.e. I don't get exactly 1 unread message for each Inbox folder, since for one Inbox I get 0 and for another I get 2. All the inboxes are on IMAP servers. The button for fixing the unreaad message count in the General tab of the BirdTray Settings dialog is disabled so I cannot try it to fix the problem.

### LINKED ISSUE #269: Incorrect unread count with Mork
**OS**: Arch Linux
**Birdtray version**: 6636ee9 (master as of 2020.02.29)
**Thunderbird version**: 68.5.0
**Birdtray origin**: From source
**Qt version**: 5.14.1-1

**Description**
Mork parser finds fake unread letters

**Expected behavior**
No unread messages in Birdtray when I have none in Thunderbird

**To Reproduce**

1. Start Birdtray
2. Check unread count in Thunderbird and Birdtray

**Additional context**
Changes in 9faac5a3 break parsing in some cases (3 errors among 40+), likely messing up with processing of scopes `0x80` and `0x9A`.

This patch reverts incorrect behavior in my case.
```patch
diff --git a/src/morkparser.cpp b/src/morkparser.cpp
index deb0d0d..cc8e882 100644
--- a/src/morkparser.cpp
+++ b/src/morkparser.cpp
@@ -552,12 +552,6 @@ inline void MorkParser::setCurrentRow( int TableScope, int TableId, int RowScope
     {
         RowScope = defaultScope_;
     }
-    
-    if (!TableId) {
-        QPair<int, int> rowMapping = rowMappings.value(abs(RowId)).value(RowScope, {0, 0});
-        TableScope = rowMapping.first;
-        TableId = rowMapping.second;
-    }
 
     if ( !TableScope )
     {
```

In addition mentioned above code has changed mork dumps for some files including `tests/res/6_Unread_Inbox.msf` and `tests/res/1_Unread_Filter.msf`.

There's one of my files, anonimized with anonymizeMork.py and in addition with timestamps changed to some nonsense.
[anon.msf.gz](https://github.com/gyunaev/birdtray/files/4272260/anon.msf.gz)

### PR DESCRIPTION
**Pull request by Abestanis**
The `unreadChildren` property in the mork file previously took precedence over the `numNewMsgs` property.
We now completly ignore the `unreadChildren` property and only rely on the `numNewMsgs` property to calculate the number of unread messages.

### Reason:
On my machine Birdtray showed 5 unread messages, but there was only a single unread message in the inbox. During debugging I noticed that there were 5 `unreadChildren` entries, each of them with the value `1`, but there was only ever one `numNewMsgs` entry.

After digging around in the Thunderbird source code, I came to the conclusion that the `unreadChildren` entry is a property of a [message thread](https://support.mozilla.org/en-US/kb/message-threading-thunderbird). (See [`kThreadUnreadChildrenColumnName`](https://github.com/mozilla/releases-comm-central/blob/master/mailnews/db/msgdb/src/nsMsgDatabase.cpp#L1447))
I'm assuming that an email can be associated with multiple threads, which is the reason the entry shows up multiple times.

The `numNewMsgs` entry on the other hand seems to be a property of a mail folder. (See [`kNumUnreadMessagesColumnName`](https://github.com/mozilla/releases-comm-central/blob/master/mailnews/db/msgdb/src/nsDBFolderInfo.cpp#L27))
I assume the mail folder would be the corresponding folder in the Thunderbird UI for the `msf` filse, e.g. `Inbox` for `INBOX.msf`.

### To discuss:
The reasoning above makes a lot of assumptions. We need to gather feedback from other users before we merge this. We should especially test this with older `msf` files. The git commit history states that you, @gyunaev, added the parsing of the `numNewMsgs` property in 322ede1 back in 2018 to support "smart folders". Can you explain what these are and what the other kind of email folders are?

This closes #217 and closes #269. Might also solve #166 and #129, but I am unable to test that.
Note that there is a different issue where Birdtray does not update the number of unread emails if they change. If anybody tests this, you can work around the issue by opening the settings and closing them with the `Ok` button, which will trigger an update.

### CODE REVIEW

**Comment by nicolamori** [comment]
Hi, I don't know if this applies, but since I'm currently having no unread message while birdtray shows 1 unread then I think that #217 might be a different issue. The one you are describing effectively multiplies the times that an unread message is counted, but there's no way to obtain 1 from 0 by multiplication.

**Comment by Abestanis** [comment]
@nicolamori I think it could be possible if you have an orphaned `unreadChildren` entry that points to a message that no longer exists, as @rkfg described in his [comment on #217](https://github.com/gyunaev/birdtray/issues/217#issuecomment-580607113). Have you tried if repairing the mail folder fixes the problem, as he suggested? Also, are you able to compile Birdtray with this patch to test if it works or do you need help with that?

**Comment by Abestanis** [comment]
@chrisgraham You mentioned in #242 ([comment](https://github.com/gyunaev/birdtray/issues/242#issuecomment-582941611)) that you see an unread counter of `2` instead of the expected `1`. Could you please verify with the debug output of Birdtray (start Birdtray with `--debug`) that the two unread counts originate from the one mail folder? Also, does the counter go to `0` if you mark the one unread message as read?

**Comment by chrisgraham** [comment]
> @chrisgraham You mentioned in #242 ([comment](https://github.com/gyunaev/birdtray/issues/242#issuecomment-582941611)) that you see an unread counter of `2` instead of the expected `1`. Could you please verify with the debug output of Birdtray (start Birdtray with `--debug`) that the two unread counts originate from the one mail folder? Also, does the counter go to `0` if you mark the one unread message as read?

I'll do that, but I need to be clear that this behaviour is not related to this commit. I was providing a few unrelated updates to my original report.

**Comment by nicolamori** [comment]
@Abestanis I found the folder which contained the "ghost" unread message by starting birdtray in debug mode, and after repairing it as you suggested the unread count was fixed.
I think I can compile the patched version by modifying the PKGBUILD, but after repairing the folder I don't experience the issue anymore. So I don't know how useful is to do it right now...

**Comment by Abestanis** [comment]
@nicolamori Good to hear that the issue is fixed for you. With this patch we would have ignored the "ghost" unread message, so I still think this was relevant to your issue.

It would still be helpful if you could test this branch. We want to get a much feedback on this as possible, but you don't need to do it if it is too much trouble. I don't think modifying the `PKGBUILD` file will do anything, but I must admit I'm not really familiar with ArchLinux. I think you need to download the sources from Github and try to follow the [build instructions](https://github.com/gyunaev/birdtray#building) (the package names might be different on ArchLinux).

**Comment by nicolamori** [comment]
@Abestanis Ok, then I will try the patch as soon as I have some spare time. The PKGBUILD is essentially a bash script that downloads, build and prepares a package, so it's sufficient to edit the location of the source code.

**Comment by nicolamori** [comment]
@Abestanis I compiled and installed commit e0c9bc15edb. Everything seems fine, but it was also with birdtray 1.7 after I repaired the folders as described in my previous posts. So is there a way to trigger the bug and see if this patched version deals with it correctly?

**Comment by gyunaev** [comment]
Smart folders required this. See https://github.com/gyunaev/birdtray/issues/17 - please test.

**Comment by nicolamori** [comment]
In this moment I got an email in a monitored inbox folder, and the unread email count didn't increase by 1 (I had no unread before and birdtray still shows no unread). Restarting birdtray with --debug gives 0 unread messages in the affected inbox. Manually marking another message in the same inbox as unread the count was raised to 2, and decreased to 1 after manually re-marking as read, so now the new unread message is correctly accounted. I'm using the patched version and nothing like this ever happened with 1.7.

**Comment by Abestanis** [comment]
@nicolamori Thank you for the report. I was able to observe a similar behaviour and was able to trace it back some problems with the mork parser (See #257).

To everybody who is testing this:
I included the fix for #257 on this branch, please recompile and check if the unread counter is still correct for your mork files.
If it doesn't work correctly, please make a copy of the mork file that doesn't work.
I will upload a Python script that you can use to remove any personal data from the mork file so you can share it here and we can examine why it doesn't work as expected.

**Comment by nicolamori** [comment]
@Abestanis I installed commit ba3e22983, but I am experiencing right now the latest problem (i.e. one unread message but nothing shown in the birdtray icon). I saved the mork files for the inbox showing the problem, I don't know exactly which one is the useful one but I have three of them. I wait for your input about what to do now.

**Comment by Abestanis** [comment]
@nicolamori Awesome! Are there three different mork files because you created a copy of the same file at different times, or is the unread counter wrong in three folders? Or are you unsure, which of the mork files that Birdtray is watching is the one that contains the unread message? In the first two cases, it would be awesome if you could upload all three of them. In the last case, let me try to help you choose the correct on:

Each mail folder in your mail account (e.g. `Inbox`, `Trash`, ...) has one file with the ending `.msf` associated to it. The name of the file tells you which folder it belongs to, e.g. `INBOX.msf` contains information about the `Inbox`, `Trash.msf` about the `Trash` folder. As you know, if you run Birdtray with the `--debug` parameter, you can get the full path to each mork file you configured Birdtray to watch. The important mork file is the one where you can see a number in parentheses next to the folder name in Thunderbird. E.g. if you see `Inbox (1)`, the `INBOX.msf` file is the one we are interested in.

Once you identified the important mork file(s), download the `anonymizeMork.py` script from #259. In order to run it, you also need Python 3 to be installed. #259 also includes a Readme file that explains how to use the script, but basically you give it the path to the mork file and it will create a mork file in the same directory with the exact same structure, but without your personal information. Make sure to check the file before you upload it, because there unfortunately isn't any guarantee that the script didn't miss anything.

If you did that, you can then upload the file(s) and post the link here. Please also specify the number of unread mails that Birdtray should have detected (its the number between the parentheses in the Thunderbird Ui) and please also include the name of the directory which contains the mail account directory (e.g. `ImapMail` when the mork file was in `.../xxxx.default/ImapMail/xxxx@yyyy.zz/INBOX.msf`).

**Comment by nicolamori** [comment]
@Abestanis I have 3 mork files because in my .thunderbird account folder I have these:
* ImapMail/my.mail.server.sbd/INBOX.msf
* ImapMail/my.mail.server.msf
* ImapMail/my.mail.server/INBOX.msf
By starting birdtray with --debug I determined that the watched file was the last one. The anonymized file is [here](https://basket.fi.infn.it:443/f/67c6deae3c/?dl=1) (BTW, the anonymizer script seems to work).

**Comment by Abestanis** [comment]
Wow, very interesting. The mork file is extremely helpful, thank you!
Have a look at these lines near the end:
```mork
<(1E7D=69242)>[-1:^9F(^AC=1)(^AD=0)(^AE=1)(^AF^82)(^B0^1E72)(^B1=12)
    (^B2=1)(^B3=0)(^B4=0)(^B5=)(^B6^1CE8)(^B7^1E7D)(^B8^88)(^A8=1)(^BB=298)
    (^A1=280)(^A6=298)(^BA^DEC)(^B9=0)(^D4=1)(^D5=1)(^D6=1e)(^D7=7d0)
    (^D8=1e)(^DA=0)(^D9=1)(^DB=0)(^88^1CD5)(^A7^82)(^A2=1)(^E5=1)(^A2=0)]
```
There are two cells for `numNewMsgs` (`A2`):
```mork
(^A2=1)...(^A2=0)
```
The first one with the correct value, the second one overwrites it with `0`. Let me do some research on how this should get parsed correctly. In the mean time, if you use the repair functionality of Thunderbird, does that fix the problem?

**Comment by Abestanis** [comment]
Ok, so tl;dr, the last entry has to be ignored.

**Comment by nicolamori** [comment]
@Abestanis it happened again just now and I tried to repair the folder, to no avail. I see that you have pushed some new commits, tell me if/when I should test a new version (and which commit, I'm used to a Gitlab workflow and I find it quite hard to follow the commit history here on Github :) ).

**Comment by Abestanis** [comment]
I was expecting the repair to remove the duplicate cell, but even better. Yes please test 56d63aa.

>  I find it quite hard to follow the commit history

Yeah, that is my fault for basing this branch on the two other pull requests and rebasing & force pushing when something changes. I should probably merge in the future.

**Comment by nicolamori** [comment]
@Abestanis Another feedback before I try commit 56d63aa: after trying to repair the folder I'm no longer able to make birdtray show any unread message for the affected inbox, regardless of how many messages I manually set as read or unread. I saved a copy of the mork file and can upload it if you want to take a look. My other inboxes work flawlessly as always.
I'll report about the new version soon.

**Comment by Abestanis** [comment]
If the new version also doesn't parse the affected mork file correctly, then please upload the copy. Please also specify what number of unread messages was shown in the Thunderbird UI at the time you made the copy.

**Comment by nicolamori** [comment]
Version 56d63aa seems to work, and marking messages as read/unread works again. I'll continue monitoring the issue but for now I have no problem.

**Comment by gyunaev** [comment]
Please test smart folders to make sure the change didn't break it.

**Comment by Abestanis** [comment]
I don't know how to create smart folders. Everything I find via Google seems to be outdated and does no longer work. The closest thing I found were `Saved Search Folders`. I tested those and they seem to work.

**Comment by gyunaev** [comment]
> I don't know how to create smart folders. Everything I find via Google seems to be outdated and does no longer work. The closest thing I found were `Saved Search Folders`. I tested those and they seem to work.

Yeah, its pretty well hidden now. See the menu View -> Folders -> Unified.

**Comment by Abestanis** [comment]
I did test the new code with the `Unified Folders` and it does work correctly, except for if the user never selected the `Unified Foder` like this:

Selecting the unified folder

If the user does not do that, the corresponding mork file contains no useful information about the unread count:

<details>
<summary>Mork file after switching to unified folders with one unread message and not selecting it.</summary>

```mork
// <!-- <mdb:mork:z v="1.4"/> -->
< <(a=c)> // (f=iso-8859-1)
  (80=ns:msg:db:row:scope:msgs:all)(81=subject)(82=sender)(83=message-id)
  (84=references)(85=recipients)(86=date)(87=size)(88=flags)(89=priority)
  (8A=label)(8B=statusOfset)(8C=numLines)(8D=ccList)(8E=bccList)
  (8F=msgThreadId)(90=threadId)(91=threadFlags)(92=threadNewestMsgDate)
  (93=children)(94=unreadChildren)(95=threadSubject)(96=msgCharSet)
  (97=ns:msg:db:table:kind:msgs)(98=ns:msg:db:table:kind:thread)
  (99=ns:msg:db:table:kind:allthreads)
  (9A=ns:msg:db:row:scope:threads:all)(9B=threadParent)(9C=threadRoot)
  (9D=msgOffset)(9E=offlineMsgSize)
  (9F=ns:msg:db:row:scope:dbfolderinfo:all)
  (A0=ns:msg:db:table:kind:dbfolderinfo)(A1=numMsgs)(A2=numNewMsgs)
  (A3=folderSize)(A4=expungedBytes)(A5=folderDate)(A6=highWaterKey)
  (A7=mailboxName)(A8=UIDValidity)(A9=totPendingMsgs)
  (AA=unreadPendingMsgs)(AB=expiredMark)(AC=version)(AD=forceReparse)
  (AE=fixedBadRefThreading)>
{1:^80 {(k^97:c)(s=9)} }
{FFFFFFFD:^9A {(k^99:c)(s=9)} }

<(80=1)(81=0)(82=Trash)>
{1:^9F {(k^A0:c)(s=9u)} 
  [1(^AC=1)(^AD=0)(^AE=1)(^A7^82)]}

@$${1{@
< <(a=c)> // (f=iso-8859-1)
  (AF=charSetOverride)(B0=charSet)>
<(83=124)>[1:^9F(^88^83)]
@$$}1}@

@$${2{@
< <(a=c)> // (f=iso-8859-1)
  (B1=searchFolderUri)(B2=searchOnline)(B3=searchStr)
  (B4=searchFolderFlag)>
<(84=ALL)(85
    =mailbox://me@localhost/Trash|mailbox://nobody@Local%20Folders/Trash)
  (86=100)>[-1:^9F(^AC=1)(^AD=0)(^AE=1)(^A7^82)(^88^83)(^B3^84)(^B1^85)
    (^B4^86)(^B2=1)]
@$$}2}@

@$${3{@
@$$}3}@
```
</details>

However, as you can see it also doesn't have any `unreadChildren` data, so the current code couldn't read the correct unread count either. As a matter of fact, I don't see any data in the mork file that indicates the unread count.

However, as soon as the user selects the `Unified Folder`, the necessary data is added to the mork file and the new code reads the unread count correctly:

<details>
<summary>Mork file after switching to unified folders with one unread message and selecting it.</summary>

```mork
// <!-- <mdb:mork:z v="1.4"/> -->
< <(a=c)> // (f=iso-8859-1)
  (B8=viewType)(B9=columnStates)(BA=MRUTime)(BB=sortColumns)
  (BC=customSortCol)(BD=folderName)(80=ns:msg:db:row:scope:msgs:all)
  (81=subject)(82=sender)(83=message-id)(84=references)(85=recipients)
  (86=date)(87=size)(88=flags)(89=priority)(8A=label)(8B=statusOfset)
  (8C=numLines)(8D=ccList)(8E=bccList)(8F=msgThreadId)(90=threadId)
  (91=threadFlags)(92=threadNewestMsgDate)(93=children)
  (94=unreadChildren)(95=threadSubject)(96=msgCharSet)
  (97=ns:msg:db:table:kind:msgs)(98=ns:msg:db:table:kind:thread)
  (99=ns:msg:db:table:kind:allthreads)
  (9A=ns:msg:db:row:scope:threads:all)(9B=threadParent)(9C=threadRoot)
  (9D=msgOffset)(9E=offlineMsgSize)
  (9F=ns:msg:db:row:scope:dbfolderinfo:all)
  (A0=ns:msg:db:table:kind:dbfolderinfo)(A1=numMsgs)(A2=numNewMsgs)
  (A3=folderSize)(A4=expungedBytes)(A5=folderDate)(A6=highWaterKey)
  (A7=mailboxName)(A8=UIDValidity)(A9=totPendingMsgs)
  (AA=unreadPendingMsgs)(AB=expiredMark)(AC=version)(AD=forceReparse)
  (AE=fixedBadRefThreading)(AF=charSetOverride)(B0=charSet)
  (B1=searchFolderUri)(B2=searchOnline)(B3=searchStr)
  (B4=searchFolderFlag)(B5=sortType)(B6=sortOrder)(B7=viewFlags)>

<(80=1)(81=0)(82=Trash)(83=124)(84=ALL)(85
    =mailbox://me@localhost/Trash|mailbox://nobody@Local%20Folders/Trash)
  (86=100)(87=1584271718)(88=12)(89=)(8B
    ={"threadCol":{"visible":true,"ordinal":"1"},"flaggedCol":{"visible":true,\
"ordinal":"3"},"attachmentCol":{"visible":true,"ordinal":"5"},"subjectCol":{"v\
isible":true,"ordinal":"7"},"unreadButtonColHeader":{"visible":true,"ordinal":\
"9"},"senderCol":{"visible":false,"ordinal":"11"},"recipientCol":{"visible":fa\
lse,"ordinal":"13"},"correspondentCol":{"visible":true,"ordinal":"15"},"junkSt\
atusCol":{"visible":true,"ordinal":"17"},"receivedCol":{"visible":false,"ordin\
al":"19"},"dateCol":{"visible":true,"ordinal":"21"},"statusCol":{"visible":fal\
se,"ordinal":"23"},"sizeCol":{"visible":false,"ordinal":"25"},"tagsCol":{"visi\
ble":false,"ordinal":"27"},"accountCol":{"visible":false,"ordinal":"29"},"prio\
rityCol":{"visible":false,"ordinal":"31"},"unreadCol":{"visible":false,"ordina\
l":"33"},"totalCol":{"visible":false,"ordinal":"35"},"locationCol":{"visible":\
true,"ordinal":"37"},"idCol":{"visible":false,"ordinal":"39"}})>
{1:^9F {(k^A0:c)(s=9)} 
  [1(^AC=1)(^AD=0)(^AE=1)(^A7^82)(^88^83)(^B3^84)(^B1^85)(^B4^86)(^B2=1)
    (^BA^87)(^B5=12)(^B6=1)(^BB=)(^B9^8B)(^A2=1)(^A1=1)]}
```
</details>

**Comment by Abestanis** [comment]
We did have feedback from multiple users and the new system seems to work. I thinkt this is now ready for your review @gyunaev.

**Comment by RYRNRY** [comment]
I am running Thunderbird v 68.3 on Windows 7.   Could someone provide link to Instructions on how to build on Windows for BIRDTRAY.

I am very much interested in trying out this fix on Windows version of BirdTray.

Thanks.

**Comment by gyunaev** [comment]
Please see the _Building_ section on the main page. Please note that it assumes you have built the Qt software on Windows before. Otherwise it might be challenging, as the initial setup is not straightforward.

**Comment by 2019-05-10** [comment]
Pulled the code yesterday, built and run -- and yet I do get the usual wrong unread count (tough only 1 today so far).
BTW: not sure if somehow related, but the button ins  Settings -> "Fix the unread message count" is never active, whatever I tried.

**Comment by Abestanis** [comment]
> Pulled the code yesterday, built and run

Did you pull the `master` branch or the old branch from this pull request? If the later is the case, please use `master`, as it also contains other fixes (such as #293). If you are already using `master`, please open a new issue, identify the mork file that generates the `1` unread count using the logging protocol window in the settings, anonymize the mork file using [this script](https://github.com/gyunaev/birdtray/tree/master/contrib/anonymizeMork) and attach the anomyzed mork file to the issue.

> BTW: not sure if somehow related, but the button ins Settings -> "Fix the unread message count" is never active, whatever I tried.

That functionality is only available to the sqlite parser and not the mork parser, but we are aware that that is not obvious.
